cmd/otel-allocator/allocation/allocator.go

```diff
@@ -88,119 +88,129 @@ func (allocator *Allocator) Collectors() map[string]*collector {
+	chosenCollector := allocator.findNextCollector()
+	targetItem := TargetItem{
+		JobName:   target.JobName,
+		Link:      LinkJSON{fmt.Sprintf("/jobs/%s/targets", url.QueryEscape(target.JobName))},
+		TargetURL: target.TargetURL,
+		Label:     target.Label,
+		Collector: chosenCollector,
+	}
+	allocator.targetItems[targetItem.hash()] = &targetItem
+	chosenCollector.NumTargets++
+	targetsPerCollector.WithLabelValues(chosenCollector.Name).Set(float64(chosenCollector.NumTargets))
+}
+
+// allCollectorsPresent checks if all the collectors provided are in the allocator's map
+func (allocator *Allocator) allCollectorsPresent(collectors []string) bool {
+	if len(collectors) != len(allocator.collectors) {
+		return false
+	}
+	for _, s := range collectors {
+		if _, ok := allocator.collectors[s]; !ok {
+			return false
+		}
+	}
+	return true
+}
+
 // SetWaitingTargets accepts a list of targets that will be used to make
 // load balancing decisions. This method should be called when there are
 // new targets discovered or existing targets are shutdown.
 func (allocator *Allocator) SetWaitingTargets(targets []TargetItem) {
+	timer := prometheus.NewTimer(timeToAssign.WithLabelValues("SetWaitingTargets"))
+	defer timer.ObserveDuration()
 	// Dump old data
 	allocator.m.Lock()
 	defer allocator.m.Unlock()
-	allocator.targetsWaiting = make(map[string]TargetItem, len(targets))
-	// Set new data
-	for _, i := range targets {
-		allocator.targetsWaiting[i.hash()] = i
+
+	// Make the temp map for access
+	tempTargetMap := make(map[string]TargetItem, len(targets))
+	for _, target := range targets {
+		tempTargetMap[target.hash()] = target
+	}
+
+	// Check for removals
+	for k, target := range allocator.targetItems {
+		// if the old target is no longer in the new list, remove it
+		if _, ok := tempTargetMap[k]; !ok {
+			allocator.collectors[target.Collector.Name].NumTargets--
+			delete(allocator.targetItems, k)
+		}
+	}
+
+	// Check for additions
+	for k, target := range tempTargetMap {
+		// Do nothing if the item is already there
+		if _, ok := allocator.targetItems[k]; ok {
+			continue
+		} else {
+			// Assign a collector to the new target
+			allocator.assignTargetToNextCollector(&target)
+		}
 	}
 }
 
 // SetCollectors sets the set of collectors with key=collectorName, value=Collector object.
 // This method is called when Collectors are added or removed.
 func (allocator *Allocator) SetCollectors(collectors []string) {
 	log := allocator.log.WithValues("component", "opentelemetry-targetallocator")
+	timer := prometheus.NewTimer(timeToAssign.WithLabelValues("SetCollectors"))
+	defer timer.ObserveDuration()
 
 	allocator.m.Lock()
 	defer allocator.m.Unlock()
 	if len(collectors) == 0 {
 		log.Info("No collector instances present")
 		return
+	} else if allocator.allCollectorsPresent(collectors) {
+		log.Info("No changes to the collectors found")
+		return
 	}
+
+	// Clear existing collectors
```

Comment:
Food for future thought: it would be ideal if we didn't need to completely reset the `allocator.collectors` map here, but only remove those that are not in `collectors` and mark their targets for reallocation.  This would avoid unnecessarily moving targets away from collectors that are still running.

Reply:
I have added suggestions for that feature.

Reply:
thank you both for the suggestion. I have implemented as much and added a test for this scenario.